paxos/proposer.hh

```diff
@@ -15,23 +15,34 @@ namespace witnesskvs::paxos {
 
 class Proposer {
  private:
+  uint64_t proposal_number_;
   uint8_t node_id_;
   int majority_threshold_;
 
   std::shared_ptr<ReplicatedLog> replicated_log_;
   std::shared_ptr<PaxosNode> paxos_node_;
+  std::vector<bool> is_prepare_needed_;
```

Comment:
I think when we become leader this vector must be initialized for all values to be true

Reply:
Or at least make the proposal number = 0

Reply:
I think we do, but let me check once.

Reply:
After a leader change we must make sure the prepares happen
>> We do, I have a unittest for this already where I change the leader and make sure we are doing prepare with new proposal id.

Reply:
I verified the proposal number 0 on leader change in both ways:
**Method 1, I have added a unittest:**

1. Start with 3 nodes, node 1 is the leader:
     It start with proposal number 0
     Gets a proposal id, after first prepare
     Uses that for it's remaining operations without prepare
2. Node 1 goes down, node 0 becomes leader:
    Node 0 start with proposal number 0
    Gets a proposal id, after first prepare
    Uses that for it's remaining operations without prepare
3. Node 1 comes backup, becomes leader:
    Node 1 start with proposal id 0
    Gets a proposal id after first prepare
    Keeps doing prepare until it fills the holes up, does 5 prepare to catch up
    Once caught up, uses the same proposal id for remaining operations without prepare
4. Node 1 goes down again, node 0 becomes the leader
   Node 0 starts with proposal id 0
   Gets a proposal id after first prepare
   Uses that for it's remaining operations without prep

All of this logic is in the unittest.

**Method 2, manual verification by running the client and server:**

1. Start 3 servers, node 1 is leader
    Same as above
2. Kill node 1, node 0 is leader now
    same as above
3. Bring back node 1, it becomes leader
    same as above, does prep with different proposal id until fills holes and then only accept with same proposal id.
4. Kill node 1 again, node 0 becomes leader
   same as above